include/gatekeeper_ggu.h

```diff
@@ -19,11 +19,66 @@
 #ifndef _GATEKEEPER_GGU_H_
 #define _GATEKEEPER_GGU_H_
 
+#include "gatekeeper_net.h"
+#include "gatekeeper_flow.h"
+
 /* Configuration for the GK-GT Unit functional block. */
 struct ggu_config {
-	unsigned int	lcore_id;
+	unsigned int      lcore_id;
+
+	/* The UDP source and destination port numbers for GGU. */
+	uint16_t          ggu_src_port;
+	uint16_t          ggu_dst_port;
+
+	/*
+	 * The fields below are for internal use.
+	 * Configuration files should not refer to them.
+	 */
+
+	/* RX queue on the back interface. */
+	uint16_t          rx_queue_back;
+	struct net_config *net;
+	struct gk_config  *gk;
+};
+
+/*
+ * Since the length of IPv6 policy is much larger than IPv4 policy,
+ * to save network bandwidth, we choose to use different data structures
+ * to store the in-packet policies for IPv4 and IPv6, respectively.
+ *
+ * Packet format: Ethernet headers + IP header + UDP header + Data.
+ * In the IP payload, the format is: num_policies (8 bits) + proto1 + proto2
+ * + ... + policy1 + policy2 + ... .
```

Comment:
Given that our goal is to pack as many policy decisions per packets as possible. The following format would save a lot of bytes when there are decline decisions:

v1, n1, n2, n3, n4: Each of these fields are 1-byte long.
v1 is a constant indicating the version of the format, in this case 1.
n1 is the number of IPv4 decline decisions.
n2 is the number of IPv6 decline decisions.
n3 is the number of IPv4 granted decisions.
n4 is the number of IPv6 granted decisions.

Field v1 will enable us to change the format, incrementally update the Gatekeeper servers, and incrementally update the Grantor servers.

Given that each one of the blocks has a known type of policy decision, each of their entries doesn't need the state field. That is, we save this one byte per entry as well. Moreover, declines are smaller than granted decisions, so we can have more declines.

Notice that this format also prioritizes declines. The reason for this is that they are likely going to have a higher impact on the bandwidth being filtered.

Opinions on this format?